File: app/src/main/java/com/blighter/algoprog/utils/UrlInterceptor.java
package com.blighter.algoprog.utils;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import com.blighter.algoprog.R;
import com.blighter.algoprog.fragments.ModuleFragment;
import com.blighter.algoprog.fragments.TaskFragment;
import com.blighter.algoprog.fragments.TaskListsFragment;

//custom WebViewClient so new materials wont trigger Main Activity intent interceptor
public class UrlInterceptor extends WebViewClient {
    Context context;

    public UrlInterceptor(Context context) {
        this.context = context;
    }

    @Override
    public boolean shouldOverrideUrlLoading(WebView view, String url) {
        if (url.contains("https://algoprog.ru/material/")) {
            String id = url.replaceAll("https://algoprog.ru/material/", "");
            if (id.contains("p")) {
                Bundle bundle = new Bundle();
                bundle.putString("id", id);
                TaskFragment taskFragment = new TaskFragment();
                taskFragment.setArguments(bundle);
                FragmentManager fragmentManager1 = ((FragmentActivity) context).getSupportFragmentManager();
                FragmentTransaction fragmentTransaction1 = fragmentManager1.beginTransaction();
                fragmentTransaction1.replace(R.id.container_in_Main, taskFragment);
                fragmentTransaction1.addToBackStack(null);
                fragmentTransaction1.commit();
            } else if (url.length() == 34) {
                Bundle bundle = new Bundle();
                bundle.putString("idForTaskList", id);
                TaskListsFragment taskListFragment = new TaskListsFragment();
                taskListFragment.setArguments(bundle);
                FragmentManager fragmentManager1 = ((FragmentActivity) context).getSupportFragmentManager();
                FragmentTransaction fragmentTransaction1 = fragmentManager1.beginTransaction();
                fragmentTransaction1.replace(R.id.container_in_Main, taskListFragment);
                fragmentTransaction1.addToBackStack(null);
                fragmentTransaction1.commit();
            } else {
                Bundle bundle = new Bundle();
                bundle.putString("url", url);
                ModuleFragment moduleFragment = new ModuleFragment();
                moduleFragment.setArguments(bundle);
                FragmentManager fragmentManager1 = ((FragmentActivity) context).getSupportFragmentManager();
                FragmentTransaction fragmentTransaction1 = fragmentManager1.beginTransaction();
                fragmentTransaction1.replace(R.id.container_in_Main, moduleFragment);
                fragmentTransaction1.addToBackStack(null);
                fragmentTransaction1.commit();
            }
        } else {
            Intent intent = new Intent(Intent.ACTION_VIEW);
            intent.setData(Uri.parse(url));
            context.startActivity(intent);
            view.reload();
        }
        return super.shouldOverrideUrlLoading(view, url);
    }
}

File: app/src/main/java/com/blighter/algoprog/utils/UrlInterceptor.kt
package com.blighter.algoprog.utils

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.Toast
import androidx.fragment.app.FragmentActivity
import com.blighter.algoprog.R
import com.blighter.algoprog.fragments.ModuleFragment
import com.blighter.algoprog.fragments.TaskFragment
import com.blighter.algoprog.fragments.TaskListsFragment
import java.io.UnsupportedEncodingException
import java.net.URLDecoder

//custom WebViewClient so new materials wont trigger Main Activity intent interceptor
class UrlInterceptor(var context: Context) : WebViewClient() {
    override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
        if (url.contains("https://algoprog.ru/material/")) {
            var id = url.replace("https://algoprog.ru/material/".toRegex(), "")
            when {
                (id.contains("p") && id.contains("^[0-9]*\$") && !id.contains(".")) -> {
                    val bundle = Bundle()
                    bundle.putString("id", id)
                    val taskFragment = TaskFragment()
                    taskFragment.arguments = bundle
                    val fragmentManager1 = (context as FragmentActivity).supportFragmentManager
                    val fragmentTransaction1 = fragmentManager1.beginTransaction()
                    fragmentTransaction1.replace(R.id.container_in_Main, taskFragment)
                    fragmentTransaction1.addToBackStack(null)
                    fragmentTransaction1.commit()
                }
                id.contains(".") -> {
                    try {
                        id = URLDecoder.decode(id, "UTF-8")
                        val bundle = Bundle()
                        bundle.putString("idForTaskList", id)
                        val taskListFragment = TaskListsFragment()
                        taskListFragment.arguments = bundle
                        val fragmentManager1 = (context as FragmentActivity).supportFragmentManager
                        val fragmentTransaction1 = fragmentManager1.beginTransaction()
                        fragmentTransaction1.replace(R.id.container_in_Main, taskListFragment)
                        fragmentTransaction1.addToBackStack(null)
                        fragmentTransaction1.commit()
                    } catch (e: UnsupportedEncodingException) {
                        Toast.makeText(context, "Произошла неизвестная ошибка", Toast.LENGTH_LONG).show()
                    }
                }
                else -> {
                    val bundle = Bundle()
                    bundle.putString("url", url)
                    val moduleFragment = ModuleFragment()
                    moduleFragment.arguments = bundle
                    val fragmentManager1 = (context as FragmentActivity).supportFragmentManager
                    val fragmentTransaction1 = fragmentManager1.beginTransaction()
                    fragmentTransaction1.replace(R.id.container_in_Main, moduleFragment)
                    fragmentTransaction1.addToBackStack(null)
                    fragmentTransaction1.commit()
                }
            }
        } else {
            val intent = Intent(Intent.ACTION_VIEW)
            intent.data = Uri.parse(url)
            context.startActivity(intent)
            view.reload()
        }
        return super.shouldOverrideUrlLoading(view, url)
    }
}